source/program/main.cpp

```diff
@@ -146,55 +146,141 @@ int multiworld_init(lua_State* L) {
             lua_call(L, 1, 1);
 
-            size_t resultSize;
             const char* luaResult = lua_tolstring(L, 1, &resultSize);
             
             if (pcallResult == 0) {
                 // success! top string is the entire result
-                output_success = true;
-                memcpy(output_start, luaResult, std::min(output_buffer_size, resultSize));
-                response = resultSize;
-            } else {
-                // error happened
-                response = snprintf(output_start, output_buffer_size, "%s", luaResult);
-            }
+                outputSuccess = true;
+            } 
+            sendBuffer->insert(sendBuffer->begin(), luaResult, luaResult + resultSize);
         } else {
-            response = snprintf(output_start, output_buffer_size, "error parsing buffer: %d", loadResult);
+            std::string errorMessage = "error parsing buffer: " + std::to_string(loadResult);
+            const char* errorAsCString = errorMessage.c_str();
+            resultSize = strlen(errorAsCString);
+            sendBuffer->insert(sendBuffer->begin(), errorAsCString, errorAsCString + resultSize);
         }
-        buffer[0] = output_success;        
-        buffer[1] = response & 0xff;
-        buffer[2] = (response >> 8)  & 0xff;
-        buffer[3] = (response >> 16) & 0xff;
-        return response + 4;
+        sendBuffer->insert(sendBuffer->begin(), PACKET_REMOTE_LUA_EXEC);
+        sendBuffer->insert(sendBuffer->begin() + 1, outputSuccess);
+        sendBuffer->insert(sendBuffer->begin() + 2, resultSize & 0xff);
+        sendBuffer->insert(sendBuffer->begin() + 3, (resultSize >> 8)  & 0xff);
+        sendBuffer->insert(sendBuffer->begin() + 4, (resultSize >> 16)  & 0xff);
+        return sendBuffer;
     });
 
     // Register calling update again
     multiworld_schedule_update(L);
     return 0;
 }
 
+std::vector<u8>* create_packet_from_lua_string(lua_State* L) {
+    size_t resultSize = 0;          // length of the lua string response (without \0)
+    std::vector<u8>* sendBuffer;  // sendBuffer to store the result. this pointer is returned 
+    
+    const char* luaResult = lua_tolstring(L, 1, &resultSize);
+
+    sendBuffer = new std::vector<u8>();
+    if (sendBuffer == NULL) return NULL;
+
+    uint32_t resultAs32Bit = resultSize;
+    u8* chars = reinterpret_cast<u8*>(&resultAs32Bit);
+    // If you move the following line by one line down the compiler throws a warning?!?
+    sendBuffer->insert(sendBuffer->begin(), luaResult, luaResult + resultAs32Bit);
+    sendBuffer->insert(sendBuffer->begin(), chars, chars + sizeof(uint32_t));
+
+    return sendBuffer;
+}
+
+/* This function gets called by the lua to sent message to the client */
+int gamelog_send(lua_State* L) {
+    if (RemoteApi::clientSubs.logging) {
+        RemoteApi::SendMessage([=]() -> std::vector<u8>* {
+            std::vector<u8>* sendBuffer = create_packet_from_lua_string(L); 

```

Comment:
If the callback is expected to be called when lua exists, then it should receive the lua_State via argument instead of capturing it.

You should be very careful to not keep the lua state around, especially when dealing with threads.

it'd be something like this:
```c++
        RemoteApi::SendMessage(L, [](lua_State* L) -> PacketBuffer {

```

Reply:
I did what you wrote but honestly and don't really get the reason. The already existing `multiworld_update`, which is for the lua execution iirc, does it the same way.

Reply:
It's coding convention when working with the Lua api: if a function interacts with Lua, it receives an state via argument

Reply:
Hmmm I orientated myself on the existing code for `ProcessCommand` , which also does it by capturing it in the callback.

See following lines (We should change that, too?)
https://github.com/randovania/exlaunch/blob/561d2b901af2a61c9169aca829b1cdd421f2d85e/source/program/main.cpp#L148-L160

Reply:
Changed that